components/batch-system/src/batch.rs

```diff
@@ -303,30 +284,50 @@ impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
                     batch.release_control(&self.router.control_box, len);
                 }
             }
-            if !batch.normals.is_empty() {
-                for (i, p) in batch.normals.iter_mut().enumerate() {
-                    let len = self.handler.handle_normal(p);
-                    batch.counters[i] += 1;
-                    if p.is_stopped() {
-                        reschedule_fsms.push((i, ReschedulePolicy::Remove));
-                    } else {
-                        if batch.counters[i] > 3 {
-                            hot_fsm_count += 1;
-                            // We should only reschedule a half of the hot regions, otherwise,
-                            // it's possible all the hot regions are fetched in a batch the
-                            // next time.
-                            if hot_fsm_count % 2 == 0 {
-                                reschedule_fsms.push((i, ReschedulePolicy::Schedule));
-                                continue;
-                            }
-                        }
-                        if let Some(l) = len {
-                            reschedule_fsms.push((i, ReschedulePolicy::Release(l)));
+
+            let mut hot_fsm_count = 0;
+            for (i, p) in batch.normals.iter_mut().enumerate() {
+                let len = self.handler.handle_normal(p);
+                batch.counters[i] += 1;
+                if p.is_stopped() {
+                    reschedule_fsms.push((i, ReschedulePolicy::Remove));
+                } else {
+                    if batch.counters[i] > 3 {
+                        hot_fsm_count += 1;
+                        // We should only reschedule a half of the hot regions, otherwise,
+                        // it's possible all the hot regions are fetched in a batch the
+                        // next time.
+                        if hot_fsm_count % 2 == 0 {
+                            reschedule_fsms.push((i, ReschedulePolicy::Schedule));
+                            continue;
                         }
                     }
+                    if let Some(l) = len {
+                        reschedule_fsms.push((i, ReschedulePolicy::Release(l)));
+                    }
+                }
+            }
+            let mut fsm_cnt = batch.normals.len();
+            while batch.normals.len() < max_batch_size {
+                if let Ok(fsm) = self.fsm_receiver.try_recv() {
+                    run = batch.push(fsm);
+                }
+                if !run || fsm_cnt >= batch.normals.len() {
+                    break;
+                }
+                let len = self.handler.handle_normal(&mut batch.normals[fsm_cnt]);
```

Comment:
How about `batch.normals.last_mut().unwrap()`?

Reply:
Ok.

Reply:
Emmm, but we still need calculate `fsm_cnt`.